harness/determined/core/_checkpoint.py

```diff
@@ -241,79 +243,124 @@ def upload(
                     "cannot call .upload(ckpt_dir=None, shard=False), which would result in doing "
                     "nothing at all"
                 )
-            return self._upload_single(ckpt_dir, metadata)
+            return self._upload_single(ckpt_dir, metadata, selector=selector)
         else:
             storage_id = None
             if self._dist.rank == 0:
                 storage_id = str(uuid.uuid4())
             storage_id = self._dist.broadcast(storage_id)
 
             assert storage_id
-            return self._upload_sharded(ckpt_dir, storage_id, metadata)
+            return self._upload_sharded(ckpt_dir, storage_id, metadata, selector=selector)
 
-    def _upload_single(self, ckpt_dir: str, metadata: Optional[Dict[str, Any]] = None) -> str:
+    def _upload_single(
+        self,
+        ckpt_dir: str,
+        metadata: Optional[Dict[str, Any]] = None,
+        *,
+        selector: Optional[Callable[[str], bool]] = None,
+    ) -> str:
         storage_id = str(uuid.uuid4())
         resources = self._storage_manager._list_directory(ckpt_dir)
 
         # Add metadata pre-upload but without counting it among resources.
         self._write_metadata_file(ckpt_dir, metadata or {})
 
-        self._storage_manager.upload(src=ckpt_dir, dst=storage_id)
+        paths: Optional[List[str]]
```

Comment:
`Optional[storage.Paths]` maybe?

Reply:
Or separately, I think you set `paths = None` and conditionally set it to something else later, I think mypy is smart enough to infer its type without an explicit type annotation here?

Reply:
> I think you set paths = None and conditionally set it to something else later, I think mypy is smart enough to infer its type

Yes, it is smart enough! Thanks for the suggestion!